### PR TITLE
GH-2107 predict the size of the data that needs to be read in DataFile

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2107 

Briefly describe the changes proposed in this PR:

In DataFile try to predict the size of the data that needs to be read so that we can get away with a single read. This improves performance because even though the operating system might be smart, it's still cheaper to get 50 bytes in one read than to get 4+20 in two reads.

---- 
PR Author Checklist: 

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by hmottestad** [comment]
I have three benchmark queries:
 - group by query: 25% faster
 - complex query: 6% faster
 - distinct predicates query: 35% faster

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java [hunk 1]
```diff
@@ -121,29 +121,61 @@ public long storeData(byte[] data) throws IOException {
 		return offset;
 	}
 
+	// This variable is used for predicting the number of bytes to read in getData(long offset). This helps us to only
+	// need to execute a single IO read instead of first one read to find the length and then one read to read the data.
+	int dataLengthApproximateAverage = 25;
```
Comment (jeenbroekstra) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java:126]: Very clever. Is 25 as a starting point based on any sort of observation of what is "typical"? Or is it more a "we need to start somewhere"?

**Comment by jeenbroekstra** [comment]
I'm curious: what made you decide to pick this up? Are you just trawling through TODO comments in the code base?

**Comment by hmottestad** [comment]
Wanting to do a bigger SHACL benchmark. Just loading 250 million triples took all night. Then querying for all distinct predicates took 1 1/2 minutes.

At least with the distinct predicates workload 50 seemed to be close to max. Since we load in 2x the estimated average I started it at 25. For the complex workload I played around with the max a bit. 200 seemed to be the sweet spot. 

I’m also looking into caching the IDFile.

**Comment by jeenbroekstra** [comment]
It's been a while since I looked at the guts of the native store but IIRC the ids are cached at the individual resource level - the NativeStoreValueFactory creates instances of IRIs and literal that cache their own  ids internally.

**Pull request merged**